include/fi_util.h

```diff
@@ -651,4 +654,43 @@ int ofi_shm_map(struct util_shm *shm, const char *name, size_t size,
 		int readonly, void **mapped);
 int ofi_shm_unmap(struct util_shm *shm);
 
+/*
+ * Name Server TODO: add support for Windows OS
+ * (osd/windows/pthread.h should be extended)
+ */
+
+enum ofi_ns_side {
+	OFI_NS_CLIENT,
+	OFI_NS_SERVER,
+}; 
+
+typedef int(*ofi_ns_service_cmp_func_t)(void *svc1, void *svc2);
+
+struct util_ns {
+	RbtHandle		ns_map;
+	pthread_t		ns_thread;
+	int			ns_port;
+	enum ofi_ns_side	ns_side;
+
+	void	*any_service_def;
+
+	size_t	name_len;
+	size_t	service_len;
+
+	ofi_ns_service_cmp_func_t	service_cmp;
+};
+
+int ofi_ns_init(int ns_port, enum ofi_ns_side ns_side,
+		size_t name_len, void *any_service_def,
+		size_t service_len,
+		ofi_ns_service_cmp_func_t service_cmp,
+		struct util_ns *ns);
+int ofi_ns_close(struct util_ns *ns);
+int ofi_ns_add_local_name(struct util_ns *ns, const char *server,
+			  void *service, void *name);
+int ofi_ns_del_local_name(struct util_ns *ns, const char *server,
+			  void *service, void *name);
```

Comment:
why have the "server" parameter for the local calls? they are supposed to talk to the local name server. "localhost" should work. To make it configuration, the information can be set as a property of the name server (in the same way as how the port is set).